File: codeforces/1900~1999/1971/h.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    if s.is_empty() {
        return Vec::new();
    }
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn not(x: i32) -> i32 {
    if x % 2 == 1 {
        x + 1
    } else {
        x - 1
    }
}

fn dfs1(node: usize, adj: &Vec<Vec<usize>>, visited: &mut Vec<bool>, ordering: &mut Vec<usize>) {
    if visited[node] {
        return;
    }
    visited[node] = true;
    for &next_node in adj[node].iter() {
        dfs1(next_node, adj, visited, ordering);
    }
    ordering.push(node);
}

fn dfs2(node: usize, rev_adj: &Vec<Vec<usize>>, components: &mut Vec<i32>, c: i32) {
    if components[node] >= 0 {
        return;
    }
    components[node] = c;
    for &next_node in rev_adj[node].iter() {
        dfs2(next_node, rev_adj, components, c);
    }
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let n = read_int::<usize>();
    let mut g: Vec<Vec<i32>> = Vec::new();
    for _i in 0..3 {
        g.push(read_vec());
    }

    let vars = n;
    let mut adj: Vec<Vec<usize>> = vec![Vec::new(); vars*2+1];
    for i in 0..n {
        let a = g[0][i].abs() * 2 - if g[0][i] < 0 { 1 } else { 0 };
        let b = g[1][i].abs() * 2 - if g[1][i] < 0 { 1 } else { 0 };
        let c = g[2][i].abs() * 2 - if g[2][i] < 0 { 1 } else { 0 };
        let not_a = not(a);
        let not_b = not(b);
        let not_c = not(c);
        adj[not_a as usize].push(b as usize);
        adj[not_a as usize].push(c as usize);
        adj[not_b as usize].push(a as usize);
        adj[not_b as usize].push(c as usize);
        adj[not_c as usize].push(a as usize);
        adj[not_c as usize].push(b as usize);
    }

    let mut visited = vec![false; vars*2 + 1];
    let mut ordering = Vec::new();
    for node in 1..=vars*2 {
        if !visited[node] {
            dfs1(node, &adj, &mut visited, &mut ordering);
        }
    }

    let mut rev_adj: Vec<Vec<usize>> = vec![Vec::new(); vars*2 + 1];
    for node in 1..=vars*2 {
        for &next_node in adj[node].iter() {
            rev_adj[next_node].push(node);
        }
    }

    ordering.reverse();
    let mut components = vec![-1; vars*2 + 1];
    let mut c = 1;
    for &node in ordering.iter() {
        if components[node] == -1 {
            dfs2(node, &rev_adj, &mut components, c);
            c += 1;
        }
    }

    for node in (2..=vars*2).step_by(2) {
        let not_node = node-1;
        if components[node] == components[not_node] {
            writeln!(writer, "NO").ok();
            return;
        }
    }
    writeln!(writer, "YES").ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}

File: codeforces/1900~1999/1971/h.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<vector<int>> g(3, vector<int>(n));
  for (int r = 0; r < 3; ++r) {
    for (int i = 0; i < n; ++i) {
      cin >> g[r][i];
    }
  }

  map<int, int> to_index;
  int idx = 0;
  for (int r = 0; r < 3; ++r) {
    for (int i = 0; i < n; ++i) {
      if (!to_index.count(g[r][i])) {
        to_index[g[r][i]] = idx++;
      }
      if (!to_index.count(-g[r][i])) {
        to_index[-g[r][i]] = idx++;
      }
    }
  }

  auto clean_up = [&](vector<vector<int>>& adj) {
    for (vector<int>& list : adj) {
      sort(list.begin(), list.end());
      list.resize(unique(list.begin(), list.end()) - list.begin());
    }
  };

  int m = to_index.size();
  vector<vector<int>> adj(m);
  auto make_2sat_clause = [&](int a, int b) {
    adj[to_index[-a]].push_back(to_index[b]);
    adj[to_index[-b]].push_back(to_index[a]);
  };

  for (int i = 0; i < n; ++i) {
    make_2sat_clause(g[0][i], g[1][i]);
    make_2sat_clause(g[1][i], g[2][i]);
    make_2sat_clause(g[2][i], g[0][i]);
  }

  clean_up(adj);

  vector<vector<int>> rev_adj(m);
  for (int node = 0; node < m; ++node) {
    for (int next_node : adj[node]) {
      rev_adj[next_node].push_back(node);
    }
  }

  clean_up(rev_adj);

  vector<bool> visited(m);
  vector<int> ordering;
  function<void(int)> dfs1 = [&](int node) {
    if (visited[node]) {
      return;
    }
    visited[node] = true;
    for (int next_node : adj[node]) {
      dfs1(next_node);
    }
    ordering.push_back(node);
  };

  for (int node = 0; node < m; ++node) {
    if (!visited[node]) {
      dfs1(node);
    }
  }

  reverse(ordering.begin(), ordering.end());

  vector<int> component(m, -1);
  function<void(int,int)> dfs2 = [&](int node, int c) {
    if (component[node] >= 0) {
      return;
    }
    component[node] = c;
    for (int next_node : rev_adj[node]) {
      dfs2(next_node, c);
    }
  };

  int c = 1;
  for (int node : ordering) {
    if (component[node] == -1) {
      dfs2(node, c++);
    }
  }

  for (int i = 1; i <= n; ++i) {
    if (!to_index.count(i) || !to_index.count(-i)) {
      continue;
    }
    int node1 = to_index[i], node2 = to_index[-i];

    if (component[node1] == component[node2]) {
      cout << "NO" << endl;
      return;
    }
  }

  cout << "YES" << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}
